### ISSUE TITLE
Terraqueous Worldgen is relatively slow

### ISSUE THREAD

**Issue by mezz** [ShetiPhian/Localizations-and-Issues#47]
I have been profiling packs in 1.11.2 to find causes of slow worldgen.
So far a lot of runaway chunk loading issues have been fixed, which were the biggest problem by far.
With all those fixed, Terraqueous is at the top of the profiling list for worldgen (even above Runic Dungeons).

I'll write what I know below, hopefully it's a good starting point:

- Clouds are slowest, it spends a lot of time in `Chunk.relightBlock` every time a block is set. I assume this is because they are semi-opaque. I don't know of a clear solution here, but it should be similar to tree leaves. I'll ask some people. _Edit:_ the simplest solution may be to make them fully transparent to light, to avoid lighting updates.

- Sticks and stones generation calls `TerrainGen.decorate` with `DecorateBiomeEvent.Decorate.EventType.TREE` which causes Biomes O' Plenty to do tree generation, making it very slow. Everything else about it is quick. It's probably best to avoid firing terrain events that look just like vanilla's.

- Flower generation has a similar problem, it calls `TerrainGen.decorate` with `DecorateBiomeEvent.Decorate.EventType.FLOWERS`, causing Biomes O' Plenty to do a lot of work.

Let me know if you need any more info!
[worldgen-profiling.zip](https://github.com/ShetiPhian/Localizations-and-Issues/files/809976/worldgen-profiling.zip)

**Comment by ShetiPhian** [comment]
Thank you for the heads up, 

I had my fair share of those runaway issues, fixed the last one a wile ago. I had forgot to reset a variable in a loop -_-

**Clouds:**
If possible I'd rather keep them semi-opaque, due to users liking the shadows they cause. (if not I'll nuke that as fixing issues is more important then visuals)
I'm currently using a modified ore generator, with the World.setBlockState flags '2' maybe '6' would be better.

**Rest:**
I didn't have the TerrainGen.decorate events originally, but added it after having issues in void/island/etc worlds.

After a few google searches and code examples this appeared to the 'the way' to know if something should or shouldn't be generated on a per chunk basis.

If my move my generation into the fired event, I'll have the same issue as BOP due to other mods triggering the event.
Not firing the event may bring back the previous issues I was having, and if another mod blocks trees mine would still generate.
What is the correct way of doing this? I have zero issues with reworking my code, I want it correct and as compatible as possible, but I'm not sure where to go from here.

**Comment by mezz** [comment]
For clouds, I'd try out the other `setBlockState` flag and see if you still get shadows, it might be a good solution.

For the decorate biome events, I think you should talk to BoP (@Adubbz) or the original author of the event (@ScottKillen) to find out the best way for mods to use it to work together.

**Comment by ShetiPhian** [comment]
I figured it would be best to profile before, change the flag, then profile after so the difference between the two could been seen. I ran into a snag though.

@mezz I think I need an idiots guide to profiling -_-

I've launched VisualVM and Minecraft. Then picked Minecraft from the applications list and clicked the profiler tab. I created a new Minecraft world and quickly jumped back to VisualVM and clicked CPU.
Waited for the world to load completely and hit the snapshot button.

I couldn't seem to find a link between my clouds and relightBlock (func_76615_h)
Figuring I must have profiled incorrectly I loaded the file you provided, and realized I just have no clue what I doing.

**Comment by mezz** [comment]
Here's where I'm looking in the profiling snapshot:
![javaw_2017-03-07_00-34-15](https://cloud.githubusercontent.com/assets/916092/23648081/f080ae4a-02cd-11e7-8d04-9fb91416d5b1.png)

First I am looking at the Server thread and following the path to find `GameRegistry.generateWorld`. Inside that, I'm checking everything to find out what mods have the slowest world generation code.

I see terraqueous's `Genbase.generate()` there as the top entry inside `GameRegistry.generateWorld`, and then I can break it down further to see where it is spending time. 

The time next to each method in the call tree is a sum of the times inside it. So the total time for all your `generate` is 7,453 ms and inside that is a breakdown adding up to that: 4,474 ms of it is spent on clouds, 1,508 ms is on Doodads, 1,209 ms on Flowers, and 261 ms on Trees.

Inside of the cloud generation function, you can see I open a few functions that take up around 99% of the time of their parent entry. Eventually I get to `Chunk.func_76615_h` (relightBlock) which is taking 4,429 ms. Since the cloud time is only 4,474 ms and all that relighting time is inside it, that means almost 100% the time in your cloud generation method is spent on relightBlock.

**Comment by ShetiPhian** [comment]
Thank you,

I was looking through the client thread, then attempted to use filters, checked 'hot spots', looked at 'show back traces' and 'find in call tree'
Finally confused myself enough to know I needed to ask for help :)

**Comment by mezz** [comment]
No problem, it definitely takes some getting used to before things make much sense.

**Comment by ShetiPhian** [comment]
I don't know.

When I profile while running through MultiMC or IntelliJ I do not have _java.lang.Thread.run()_ under 'Server thread'
I have _java.util.concurrent.FutureTask.run()_, _net.minecraftforge.common.chunkio.ChunkIOProvider.run()_, and _net.minecraft.server.management.PlayerChunkMapEntry$1.run()_

Also the only refference to relightBlock / func_76615_h is found under Client thread

When I attempt to profile via the old or new Mojang launcher, it seems to be going normally but both Minecraft and VisualVM drop down to 0% CPU usage and sit there until I end the task 30min later.

**Comment by ShetiPhian** [comment]
Thank you, 

Where my generators hit 7,400+ they are now 102 or less at world creation, and often drop off the list completely when new chunks start generating. (Testing with the 'All The Mods 2' pack)

It appears my changes also adjusted BOP's numbers. It dropped from above generateWorld() to bellow it.
Must be related to the extra work firing the event three addition time a chunk caused.

Clouds:
Setting the flag to 6 did not do anything. Using an opacity of 0 did fix the generator but messed the lighting in cloud homes that had a mix of generated blocks and user placed blocks.
I worked around the issue by mimicking Chunk.setBlockState and not calling any lighting code.

Rest:
I've subscribed to the decorate event, with the lowest priority, and cache the result.
When my generators run they check if there is a result cached and use it. If no cache exists I fire the event and cache the result for the other generators.
